### PR TITLE
[WIP] A Deep Reinforced Model for Abstractive Summarization

### PR DESCRIPTION
**Pull request by pltrdy**
**Update:** I made a post dedicated to this implementation: http://forum.opennmt.net/t/about-a-deep-reinforced-model-for-abstractive-summarization/1347

---
Working to implement [Paulus et al (2017), A Deep Reinforced Model for Abstractive Summarization](https://arxiv.org/abs/1705.04304).

## What's in the model (and what is implemented):

- [x] temporal attention on source
   - attention between decoding `h` and encoder outputs
   - normalizing w.r.t. previous decoding step => penalizing previous high attention scores
- [x] intra-decoder attention
   - attention between `hd_t` and `[hd_1; hd_t-1] (for t>0)`
- [x] pointer-generator
   - [x] using a soft-switch 
   - [x] using `tgt_embedding` i.e. `pointer_generator.linear = proj(tgt_embedding)` (helps learning by using semantic informations)(requires partially shared embeddings)
- [x] Reducing Exposure Bias
   - [x] output predictions at each decoding steps
   - [x] using predicted token with 0.25 probability
- [x] partially sharing embeddings
   - [x] sharing tgt & src embeddings while having different vocabulary size
    `src_vocab > tgt_vocab` ; `src_embedding = linear( tgt_embedding )`
- [x] Reinforcement Policy
   - [x] Using ROUGE score to adjust learning

## Implementations
- Most of my edits are in `onmt.Reinforced`
- I added the `-reinforced` flag to use this model. 
- I tried to used as much existing functions as possible
- note that, because we need prediction at each time step the training process is quite different (run `.backward()` at each time steps. This is why I use a slightly modified `Trainer` class. It seems to have a bad impact on performance tho, GPU load < 100% most of the time.
- I currently use a different generator that is, in his current version, just using the CopyGenerator. I will be then upgraded in order to use the embedding matrix as part of its linear projection 

## Discussion
- I was first testing on really simple case like sequence copy i.e. using `src = tgt = ptb.train.txt` the model learns it really easily (more than >90% in 1 epoch on PTB). I specifically tested the copy mechanims by setting very low vocab size and looking at translation, which is correct even with a bad source (lots of unk).

- I am still finding lots of repetitions even on copy tasks. The model may continue to generate the last word for some time i.e. probably don't manage to generate `<eos>` for some reason.

- I am currently investigating it, I ran some not very promising training on summarization task (CNN/DM). The ppl goes as low as 30 (and ~30% acc) but translation give really poor results (lots of repetitions etc) there may be some issues about it. 

## What's next
- I would like to reproduce the "ML part" (without RL) result of the paper before moving to Reinforcement learning.
- Small implementation improvement like partial embeddings. 

If people here are interested by this, running tests and suggesting improvement would help a lot :)

### CODE REVIEW

**Comment by srush** [comment]
Do you have any preliminary numbers for this model?

**Comment by pltrdy** [comment]
Not at the moment. I'm training a model but I've serious doubts about how it may perform. I'm investigating.

**Comment by mataney** [comment]
Hey man, This is some exciting work!
I would love taking part of this.
How can I join?

**Comment by pltrdy** [comment]
@mataney nice to hear! :) 
The model still have issues to learn real problems i.e. my goal is to work on summarziation using CNN/DM dataset. The current version does not manage to learn interesting stuff on it. 

If you are interested in experimental work we could both run experiment on this model on various tasks and share our results in order to interpret it and patch/improve the model.

I quite like using Gitter to communicate (https://gitter.im/pltrdy) that uses your github account. I'll update my repository with my last work and give you a more precise picture of my current problems, what I tried, etc

**Comment by srush** [comment]
Hey just wanted to check in if this work was still ongoing.

**Comment by pltrdy** [comment]
It is ongoing. I'm mostly working in a separate branch to run experiments. I get some motivating results, but still quite far from the published results.

**Comment by ScarletPan** [comment]
@pltrdy Does it work now? I'm trying to reproduce their results but also failed.

**Comment by pltrdy** [comment]
@ScarletPan sry I missed this notif. 
I am still working on it outside of this PR as it wouldn't be easy to collaborate currently. I did changes that broke compatibility, thus couldn't merge. There will be an important work to merge, cleanup and refactor the current code, but if I get interesting results I will. 

Have you tried to reproduce using my code? Your own code based on onmt-py? your own code from scratch? We could share about it. 

There's still one small thing to implement in order to get better results then I'll start the merging process, and then probably publish a post to present it (implementation, results, how to reproduce, various discussions etc).

**Comment by pltrdy** [comment]
I achieved kind of good scores recently:
* **ROUGE-1:** 37.444
* **ROUGE-2:** 16.184
* **ROUGE-L:** 34.642

Some work has been done on another branch. There would be a ton of work to just resolve conflict to be honest, but I may write a guide to reproduce.

**Comment by pltrdy** [comment]
@srush @sebastianGehrmann are we going to merge this one?

Note that I initially implemented an 'avoid_trigram_repetition' behavior that is similar to your 'block_ngram_repeat'. I merge with both (one should be removed) as I think it could be interesting to compare performance. I'm using Tensor operations which may be faster (it's only considering trigram repetition but it could be easily generalized).

**Comment by pltrdy** [comment]
Hmm, I need to investigate on those errors..

**Comment by mojesty** [comment]
@pltrdy as an PR is still open I write here. I would like to implement RL strategy for BLEU-like metric (specifically, GLEU, as my objective is to fix grammatical errors), so as I don't want to reinvent the wheel, I would like to use your code. However, I have some questions about its structure
1) Why not to use Beam Search while performing RL training? You can get two beautiful things right out of the box: i) generation of multiple hypotheses as it is useful for RL and ii) get rid of code deduplication. It is still backward compatible with MLE training if you make slight changes 
2) I was unable to find all the classes that you've described in the blog post, specifically, RTrainer

Sadly this PR has not been merged yet, as my task is just simple seq2seq may be it is easier to reimplement things while using mostly already wriiten code.

**Comment by pltrdy** [comment]
Hi @mojesty,

1. that could be done. I have just been following Paulus 2017 paper which performs greedy search during training. From what I discussed with them, I think they didn't tried beam search at training time. 
2. The code has been refactored multiple times, in order to organise it better: put modules in ./onmt/modules avoid code duplicates etc. RTrainer was a voluntary code duplicate in order to keep things simple during development. The training logic is in the "normal" trainer.

**Comment by pschwllr** [comment]
Hi @pltrdy,

I was just looking at your drmas branch. In Reinforced.py, where you define the mixed loss, there is one bracket that is put at the wrong place. Correct me if I am wrong, but, I guess, line 332

`loss = (self.gamma * rl_loss) - ((1 - self.gamma * ml_loss))`

should be

`loss = (self.gamma * rl_loss) - ((1 - self.gamma) * ml_loss)`.

Otherwise, great work! 

Have you planned to make it compatible with the OpenNMT pytorch0.4 only version?

**Comment by pltrdy** [comment]
@pschwllr Wow, that's a pretty big thing. I'm not sure how I even got descent results with it.
Did you trained before/after fixing it?

**Comment by pschwllr** [comment]
@pltrdy Yeah, quite surprising that you could get any results with this loss!

I was interested in seeing some OpenNMT RL example code, that's why I skimmed through it. 

I didn't retrain it for abstract summarisation, but would be interesting to compare the results before/after changing the loss.

**Comment by mojesty** [comment]
@pltrdy Could you please provide a pre-trained model? Due to the limited computational resources, I cannot train it myself.

**Comment by pltrdy** [comment]
To give some background, I've established interesting results in late February that I presented in the OpenNMT workshop, announcing (F1 scores on CNN/DM): 

| Rouge-1 | Rouge-2 | Rouge-L |
|-------------|-------------|--------------|
| 38.27      | 16.57     | 35.29       |

At the time, I was working on another -- even more experimental (understand: dirty) -- branch [`drmas_wip`](https://github.com/pltrdy/OpenNMT-py/tree/drmas_wip). 

Then, I did cleanup and merging, and never got those results back. 

So here's where we are:
- this PR is attached to `drmas` branch, which is closer in term of code different/conflicts, but does not provide good results.
- best results mights be found using `drmas_wip`, but this is an old code, that will require `torch=0.3.1` and `torchtext=0.1.1` (I've a separate cuda environment for that).

Here's what needs to be done:

1. validate results on `drmas_wip` (confirm that the branch is still a state that produce such results)
2. try to reproduce on `drmas`
3. resolve conflict or open a new PR (may be easier).

I will - short term (~= 1 week) run point 1. and 2. and if good results are found, try to 3.

Feel free to experiment with the code and be critical about it, I'm sure that it's far from perfect.

**Comment by elisemicho** [comment]
@pltrdy May I ask what is the state of the code and results at the moment?

**Comment by rylanchiu** [comment]
Hi @pltrdy , what is the version of pytorch with this implementation? `0.4` or `0.3` ? Thanks!

**Comment by pltrdy** [comment]
@elisemicho I cannot reproduce. It would require - not trivial - debugging. I've not much time, therefore, it's kinda nowhere :/ 

@rylanchiu `torch=0.3.1` and `torchtext=0.1.1`

**Comment by qianyuwang** [comment]
@pltrdy Hi.  It's OK in training process using your codes. But I have problem in translation process.  The problem is as below.
image
The codes:
image

The problem is trace back at line 329 in 'Translator.py'.  After printing the variables. I think the variable 'batch' changed after going through line 321. I can't get 'batch.indices'.  What happened in the decoder forward. Do you have any idea?

**Comment by pltrdy** [comment]
@qianyuwang which branch do you use? (the "original" was `drmas_wip`. `drmas` is a bit more clean, I tried to keep it up to date, best results are supposed to be on the older, more experimental `drmas_wip`).

Did you pass `-reinforced` at translation time? It may be use (need to check it as well).

Let's discuss it on Gitter (you can log with Github): https://gitter.im/onmtpy-drmas/Lobby

**Comment by du-david** [comment]
Hi @pltrdy,

I wonder if the rl_loss is computed correctly. I printed out the value of the variable metric, i.e. r(y_hat)-r(y_s), and the value of rl_loss after rl_loss=(rl_loss*metric).sum(). I saw that when metric is positive, rl_loss is also positive.

I think it is wrong. The ml_loss is negative log likelihood so always positive, but the rl_loss is metric times positive log likelihood. If metric is positive, rl_loss should be negative, and vice versa. In your implementation, the rl_loss is firstly computed in the decoder, so I think the decoder returns, like the ml_loss, the negative log likelihood of the sampled sequences. The line for combining with the metric should be
rl_loss=(-1.0*rl_loss*metric).sum()
Am I wrong?

I also wonder if you have the copy or pointer-generator integrated. I don't find them in the code.

Thanks!

**Comment by pltrdy** [comment]
In the paper:

> We can see that minimizing `L_rl` is equivalent to maximizing the conditional likelihood of the sampled
> sequence `y_s` if it obtains a higher reward than the baseline `ŷ`, thus increasing the reward expectation of our model.

`rl_loss` precisely tries to avoid  `metric > 0`, because it means that `r(ŷ) > r(y_s)`. Thus, it leads to positive loss.

**Note** that their's a missing ")" in `loss = (self.gamma * rl_loss) - ((1 - self.gamma * loss))` as suggested in https://github.com/OpenNMT/OpenNMT-py/pull/319#issuecomment-399969575

**Comment by du-david** [comment]
The line of loss is a typo which I saw myself, but the line of rl_loss is not that obvious.

I don't think rl_loss tries to avoid metric>0. y_s is sampled randomly, so you get good and bad sequences. When metric<0, rl_loss>0 and you reduce rl_loss to increase the probability of y_s being generated. When metric>0, rl_loss<0 and you also reduce rl_loss, because y_s has a worse reward and you want to decrease its probability. This is exactly the idea of RL, encourage actions that lead to high reward and discourage actions that lead to low reward. So I think a negative sign is missing from rl_loss.

I tested your code with different gamma values from 0.0 to 1.0, but ran into a cuda error in the copy generator line 373 with gamma=1.0 or even 0.9999, I got results with gamma=0.99. I wonder if you tried gamma=1.0 and if you know how to fix it.

I also tried replacing the pointer generator by the copy generator and saw no differences in the results. Can you tell me what the differences are between them and why you multiple model_opt.rnn_size by 3?

**Comment by pltrdy** [comment]
That's not how I understand what's in the paper. Moreover, I don't get what, in the paper, would justify addin "1 - [...]". However, I'm not too confident with my implementation since I had really hard time working on it, and it's not even working atm.

I've not many experiment with current code (with fixed loss). My trainings are failing when RL comes in.

Concerning the PointerGenerator: 
- The `PointerGenerator` defined by Paulus uses the concatenation of 3 vectors of size `rnn_size`:
   - encoder-context the decoder state `hd_t`, 
   - encoder-context `ce_t` (from intra encoder attention), 
   - decoder-context `cd_t` (from intra decoder attention).  
- Also, the linear projection of the `PointerGenerator` shares weights with the embedding (sect. 2.4), which lower the number of parameters, and allow the generator to make use of semantic knowledge.

Do not hesitate to experiment and share your progress. I don't have much time to experiment it at the moment but I'll be happy to help you if I can. As a side note, the same team did a recent paper ([ Kryściński, et al, 2018) in which they extend their model. That would be another good reason to get this working.

----
* [Paulus et al (2017): https://arxiv.org/abs/1705.04304](https://arxiv.org/abs/1705.04304)
* [Kryściński et al (2018): https://arxiv.org/abs/1808.07913](https://arxiv.org/abs/1808.07913)

**Comment by du-david** [comment]
Thanks pltrdy, but I don't understood "what would justify adding 1-[]". Probably I wasn't clear.

In your code rl_loss=(metric*rl_loss).sum() looks just like eq 15, but rl_loss on the right hand side is actually computed in the decoder, so like ml_loss, it is negative log likelihood. In eq 15, it is metric times positive log likelihood. That is why I think a negative sign is missing in the code.

Anyway, it is a nice implementation, although I did fix a few small things to get it run. On my problem, a non-zero gamma made the results worse by a little and the random sampling had some impacts, but my dataset is very small. I will try it on summarization if I have time. I do have a feeling that this rl strategy is a bit simple. For example, it is very hard for random sampling to beat greedy search on training data and whether sampling one sequence per training data is enough. Thanks for telling me the other paper.

**Comment by pltrdy** [comment]
Alright, I now get your point, it definitely make sense, thanks. 

I'm curious about which commit you choose as the starting point: as I said there's 2 branches on my repo related to this, namely, `drmas` and `drmas_wip` (naming isn't *that* relevant I must admit).

You can open PR to my repo, which would be appreciated, at least to discuss/share what fixes you found relevant. I previously found that such models are quite fast to converge, I'll probably run some experiments including you suggestions (from `drmas_wip` on CNN/DM)

**Comment by du-david** [comment]
I used drmas because I read that drmas_wip required an older version of pytorch. I have no problem with gamma in [0,0.99]. In [0.999,1.0] I ran into a cuda error in the copy generator, but the models were trained successfully and I can use them for inference.

On a dataset of 17k sentences, gamma=0 gave the best results, the same as with opennmt. For 0<gamma<=0.99, the results are okay but vary largely due to random sampling. For gamma>0.999, the results are very bad, gamma=1.0 gave random outputs with 0 score. My dataset is too small, and we should try to repeat the results on the datasets from the paper.

I will PR next week, but it is what you did with a few small changes. There is another problem. During training, there were some 0-length sentences and they were used for training. Did you see this in your experiments?

**Comment by vince62s** [comment]
RIP

**Comment by vince62s** [comment]
actually in the light of recent RLHF developments @pltrdy you were such a visionary :) want to try again ?

**Comment by xcfcode** [comment]
> RIP

Why RIP???

**Comment by pltrdy** [comment]
> actually in the light of recent RLHF developments @pltrdy you were such a visionary :) want to try again ?

Yep it has become a really interesting field! Not much time to allocate on this however :/

**Comment by xcfcode** [comment]
@pltrdy sorry for misunderstanding!
